### PR TITLE
Green skill classifier

### LINKED ISSUE #45: Experiment with skills mapping pipeline

### PR DESCRIPTION
**Pull request by lizgzil**
---

# Description

Improves the way we map skills to ESCO green skills by training a classifier.

- Adds a green skills classifier [green_skill_classifier.py](https://github.com/nestauk/dap_prinz_green_jobs/pull/67/files#diff-5b1eef6538de374e362539cc95b1697a5a54742f4027da22d0d1924edbb63ded)
- Creates a new function `.get_measures` in SkillMeasures which inputs the job advert. Before this wasn't actually in a function, and there were 2 instances of repeating code to do it.
- Changes to [/skill_measures_utils.py](https://github.com/nestauk/dap_prinz_green_jobs/pull/67/files#diff-2698988567d6467a6a8bdacb5371d35dc1de9a40fe6c86519e13f0108b3eab39) to include the green skill classifier
- Updates README
- Adds a util function [map_skills_utils.py](https://github.com/nestauk/dap_prinz_green_jobs/pull/67/files#diff-482143cae871314d57fc55468f4bd1b25300a0e4c8ae6b2c223ef8163a47c907) to map the skills to all of ESCO (not just green ESCO). This is comprised of code copied from ojd_daps_skills (I've tried to make it clear what is copied). So doesn't really need reviewing. Its a temporary fix, but ideally in the future we'd change ojd_daps_skills to have this functionality - unless it turns out we never use the skills mapped to all of ESCO
- skill measures test [test_skill_measures.py](https://github.com/nestauk/dap_prinz_green_jobs/pull/67/files#diff-133a94cdcb82daaa411ff73396cd4e4eea7761d6cf56fe2161c74879e7b5e04b)
- New get_embeddings function in [bert_vectorizer.py](https://github.com/nestauk/dap_prinz_green_jobs/pull/67/files#diff-5f769ae27b765d9eb7317ad265da6c7d55606be3815cc3afcff584dc6d3d70f4) - just because this code kept being repeated

Fixes #45 #27  #10 (since it uses ONET green topics in the classifier)

# Instructions for Reviewer

In order to test the code in this PR you need to ...

See if the tests work:
```
pytest
```

Check out the new green skill classifier:
```
from dap_prinz_green_jobs.pipeline.green_measures.skills.green_skill_classifier import GreenSkillClassifier

green_skills_classifier = GreenSkillClassifier()
green_skills_classifier.load_esco_data()
green_skills_classifier.load(
    model_file="s3://prinz-green-jobs/outputs/models/green_skill_classifier/green_skill_classifier_20230906.joblib"
)

pred_green_skill = green_skills_classifier.predict(
    ["Excel skills", "Heat pump installation skills"]
)

>>> [('not_green', 1.0, ('carry out sample analysis', '82423b5c-486f-42e7-b00e-7358757a8de5', 0.2772792296638087)), ('green', 0.7191159533073931, ('heat pump installation', '00735755-adc6-4ea0-b034-b8caff339c9f', 0.9072619656040537))]
```

Check out the new green skills measures

```
from dap_prinz_green_jobs.pipeline.green_measures.skills.skill_measures_utils import (
    SkillMeasures,
)
sm = SkillMeasures(config_name="extract_green_skills_esco")
sm.initiate_extract_skills(local=False, verbose=True)

job_adverts = [
    {"id": 1, "job_text": "This job requires communication and Excel skills. We want someone with experience in sustainability. Heat pump installation skills would be useful. We have a pension scheme benefit."},
    {"id": 55, "job_text": "This job contains no skills."},
    {"id": 3, "job_text": "This job contains Excel skills."},
    {"id": 8, "job_text": "This job contains a really long sentence. Promote good practice of material sustainability (reuse and or recycle) initiatives to reduce waste and save costs."},
    ]

# We will load the green taxonomy embeddings from S3 since they have already been calculated
taxonomy_skills_embeddings_dict = sm.get_green_taxonomy_embeddings(
    output_path="outputs/data/green_skill_lists/green_esco_embeddings_20230815.json", load=True)

prop_green_skills = sm.get_measures(job_adverts)
```
# Next:

After applying these changes to the mixed ojo sample, the new results for these green skills will be in `outputs/data/ojo_application/extracted_green_measures/20230912/ojo_sample_skills_green_measures_production_False_base.json`

I intend to rerun some of the analysis using these/quality check the most common green skills etc.

# Checklist:

- [ ] I have refactored my code out from `notebooks/`
- [ ] I have checked the code runs
- [ ] I have tested the code
- [ ] I have run `pre-commit` and addressed any issues not automatically fixed
- [ ] I have merged any new changes from `dev`
- [ ] I have documented the code
  - [ ] Major functions have docstrings
  - [ ] Appropriate information has been added to `README`s
- [ ] I have explained this PR above
- [ ] I have requested a code review

### CODE REVIEW

File: dap_prinz_green_jobs/pipeline/green_measures/skills/green_esco_formatting.py [hunk 1]
```diff
@@ -31,11 +31,9 @@
         "escoe_extension/outputs/data/skill_ner_mapping/esco_data_formatted.csv",
     )
 
-    formatted_green_esco_skills = (
-        formatted_esco_skills[formatted_esco_skills["id"].isin(esco_green_skills["id"])]
-        .reset_index(drop=True)
-        .set_index("id")
```
Comment (lizgzil) [inline dap_prinz_green_jobs/pipeline/green_measures/skills/green_esco_formatting.py:37]: this was causing the output file to not contain the ESCO ID

File: dap_prinz_green_jobs/pipeline/green_measures/green_measures.py [hunk 2]
```diff
@@ -105,7 +119,10 @@ def __init__(
         self.im.load_ch()
 
         # Skills attributes
-        self.sm = SkillMeasures(config_name="extract_green_skills_esco")
+        self.sm = SkillMeasures(
+            config_name="extract_green_skills_esco",
+            green_skills_classifier_model_file=self.green_skills_classifier_model_file,
```
Comment (india-kerle) [inline dap_prinz_green_jobs/pipeline/green_measures/green_measures.py]: perhaps `green_skills_classifier_model_file_name` to be as clear as possible?

File: dap_prinz_green_jobs/pipeline/green_measures/green_measures.py [hunk 3]
```diff
@@ -115,39 +132,22 @@ def get_skill_measures(
...
-        )
-
         taxonomy_skills_embeddings_dict = self.sm.get_green_taxonomy_embeddings(
             output_path=self.green_tax_embedding_path,
             load=self.load_taxonomy_embeddings,
         )
 
-        all_extracted_green_skills_dict = self.sm.map_green_skills()
-
         prop_green_skills = self.sm.get_measures(
```
Comment (india-kerle) [inline dap_prinz_green_jobs/pipeline/green_measures/green_measures.py:140]: nice - this is really clean

File: dap_prinz_green_jobs/pipeline/green_measures/skills/README.md [hunk 4]
```diff
@@ -30,60 +30,97 @@ To add the custom config file, formatted skills taxonomy and taxonomy embeddings
...
+
+This dataset contains 971 skills labelled as not-green and 743 labelled as green.
+
+The most recently saved model `s3://prinz-green-jobs/outputs/models/green_skill_classifier/green_skill_classifier_20230906.joblib` has the following test metrics:
+
+```
+              precision    recall  f1-score   support
+
+       green       0.90      0.86      0.88       175
+   not_green       0.91      0.94      0.92       254
```
Comment (india-kerle) [inline dap_prinz_green_jobs/pipeline/green_measures/skills/README.md:52]: gorgeous!

File: dap_prinz_green_jobs/pipeline/green_measures/skills/README.md [hunk 5]
```diff
@@ -30,60 +30,97 @@ To add the custom config file, formatted skills taxonomy and taxonomy embeddings
...
+This trained model can be loaded and used by running:
+
+```python
+from dap_prinz_green_jobs.pipeline.green_measures.skills.green_skill_classifier import GreenSkillClassifier
+
+green_skills_classifier = GreenSkillClassifier()
+green_skills_classifier.load_esco_data()
+green_skills_classifier.load(
+    model_file="s3://prinz-green-jobs/outputs/models/green_skill_classifier/green_skill_classifier_20230906.joblib"
+)
```
Comment (india-kerle) [inline dap_prinz_green_jobs/pipeline/green_measures/skills/README.md:69]: I initially had issues with loading the model - i think this had to do with different scikit learn versions - i had to uninstall the ojo-dap-skills library first, then re-install the requirements.

File: dap_prinz_green_jobs/pipeline/green_measures/skills/README.md [hunk 6]
```diff
@@ -30,60 +30,97 @@ To add the custom config file, formatted skills taxonomy and taxonomy embeddings
...
+
+pred_green_skill = green_skills_classifier.predict(
+    ["Excel skills", "Heat pump installation skills"]
+)
+
+>>> [('not_green', 1.0, ('carry out sample analysis', '82423b5c-486f-42e7-b00e-7358757a8de5', 0.2772792296638087)), ('green', 0.7191159533073931, ('heat pump installation', '00735755-adc6-4ea0-b034-b8caff339c9f', 0.9072619656040537))]
+
```
Comment (india-kerle) [inline dap_prinz_green_jobs/pipeline/green_measures/skills/README.md:76]: This is what I got when I ran this code!

File: dap_prinz_green_jobs/pipeline/green_measures/skills/README.md [hunk 7]
```diff
@@ -30,60 +30,97 @@ To add the custom config file, formatted skills taxonomy and taxonomy embeddings
...
 
 Note: predicting skill entities and embedding them for comparison with the green skills taxonomy can take a long time if you are inputting many job adverts.
 
-```
-from dap_prinz_green_jobs.pipeline.green_measures.skills.skill_measures_utils import SkillMeasures
+```python
 
+from dap_prinz_green_jobs.pipeline.green_measures.skills.skill_measures_utils import (
+    SkillMeasures,
+)
```
Comment (india-kerle) [inline dap_prinz_green_jobs/pipeline/green_measures/skills/README.md:103]: I was also able to run this once I re-installed requirements in my environment!

File: dap_prinz_green_jobs/pipeline/green_measures/skills/green_esco_formatting.py (hunk 1 above)
Comment (india-kerle) [reply dap_prinz_green_jobs/pipeline/green_measures/skills/green_esco_formatting.py:37]: good spot!

File: dap_prinz_green_jobs/pipeline/green_measures/skills/green_skill_classifier.py [hunk 8]
```diff
@@ -0,0 +1,429 @@
...
+from sklearn.preprocessing import StandardScaler
+from sklearn.ensemble import RandomForestClassifier
+from sklearn.metrics import classification_report
+from sklearn.metrics.pairwise import cosine_similarity
+import spacy
+
+from dap_prinz_green_jobs.utils.bert_vectorizer import get_embeddings
+from dap_prinz_green_jobs.getters.data_getters import load_s3_data, save_to_s3
+from dap_prinz_green_jobs.getters.skill_getters import get_green_skills_taxonomy
+from dap_prinz_green_jobs import BUCKET_NAME, logger
```
Comment (india-kerle) [inline dap_prinz_green_jobs/pipeline/green_measures/skills/green_skill_classifier.py]:
```suggestion
from dap_prinz_green_jobs import BUCKET_NAME, OJO_BUCKET_NAME, logger
```

File: dap_prinz_green_jobs/pipeline/green_measures/skills/green_skill_classifier.py [hunk 9]
```diff
@@ -0,0 +1,429 @@
...
+)
+
+from tqdm import tqdm
+from typing import List, Union, Tuple
+import joblib
+import s3fs
+from datetime import datetime
+import os
+
+OJO_BUCKET_NAME = "open-jobs-lake"
```
Comment (india-kerle) [inline dap_prinz_green_jobs/pipeline/green_measures/skills/green_skill_classifier.py]:
```suggestion
```

File: dap_prinz_green_jobs/pipeline/green_measures/skills/green_skill_classifier.py (hunk 8 above)
Comment (india-kerle) [reply dap_prinz_green_jobs/pipeline/green_measures/skills/green_skill_classifier.py]: This exists in the init file so you can just:

`from dap_prinz_green_jobs import OJO_BUCKET_NAME`

File: dap_prinz_green_jobs/pipeline/green_measures/skills/green_skill_classifier.py [hunk 10]
```diff
@@ -0,0 +1,429 @@
...
+    """
+    Find the green topics which are contained within a skill entity using exact matching
+    Args:
+        skill_ent: A skill entity
+        green_topics: A list of green topics
+    Returns:
+        list: A list of the green topics found in this skill entity (can be empty)
+    Example:
+        find_green_topics("We want someone with an interest in sustainability", ["Sustainability", "Green Energy"])
+        returns ["Sustainability"]
```
Comment (india-kerle) [inline dap_prinz_green_jobs/pipeline/green_measures/skills/green_skill_classifier.py:44]: this is helpful!

File: dap_prinz_green_jobs/pipeline/green_measures/skills/green_skill_classifier.py [hunk 11]
```diff
@@ -0,0 +1,429 @@
...
+from typing import List, Union, Tuple
+import joblib
+import s3fs
+from datetime import datetime
+import os
+
+OJO_BUCKET_NAME = "open-jobs-lake"
+
+
+def find_green_topics(skill_ent: str, green_topics: list) -> list:
```
Comment (india-kerle) [inline dap_prinz_green_jobs/pipeline/green_measures/skills/green_skill_classifier.py]:
```suggestion
def find_green_topics(skill_ent: str, green_topics: List[str]) -> Union[List[str], List[None]]:
```

File: dap_prinz_green_jobs/pipeline/green_measures/skills/green_skill_classifier.py [hunk 12]
```diff
@@ -0,0 +1,429 @@
...
+        # Basic cleaning
+        green_topic = green_topic.lower()
+        clean_skill_ent = skill_ent.lower()
+        clean_skill_ent = " ".join(clean_skill_ent.split())
+        if green_topic in clean_skill_ent:
+            found_green_topics.append(green_topic)
+    return found_green_topics
+
+
+def process_green_topic_data(nlp):
```
Comment (india-kerle) [inline dap_prinz_green_jobs/pipeline/green_measures/skills/green_skill_classifier.py]:
```suggestion
def process_green_topic_data(nlp) -> List[str]:
```

File: dap_prinz_green_jobs/pipeline/green_measures/skills/green_skill_classifier.py [hunk 13]
```diff
@@ -0,0 +1,429 @@
...
+
+    enhanced_green_topics = list(enhanced_green_topics)
+
+    return enhanced_green_topics
+
+
+def get_closest_match(
+    skills_list_embeddings_dict: dict,
+    green_data_embeddings_dict: dict,
+    formatted_green_data: pd.DataFrame(),
```
Comment (india-kerle) [inline dap_prinz_green_jobs/pipeline/green_measures/skills/green_skill_classifier.py]:
```suggestion
    formatted_green_data: pd.DataFrame,
```

File: dap_prinz_green_jobs/pipeline/green_measures/skills/green_skill_classifier.py [hunk 14]
```diff
@@ -0,0 +1,429 @@
...
+    )  # 'wastewater management , treatment , and reduction'
+
+    enhanced_green_topics = list(enhanced_green_topics)
+
+    return enhanced_green_topics
+
+
+def get_closest_match(
+    skills_list_embeddings_dict: dict,
+    green_data_embeddings_dict: dict,
```
Comment (india-kerle) [inline dap_prinz_green_jobs/pipeline/green_measures/skills/green_skill_classifier.py]: could use typing hints here also

File: dap_prinz_green_jobs/pipeline/green_measures/skills/green_skill_classifier.py [hunk 15]
```diff
@@ -0,0 +1,429 @@
...
+        skill_threshold=0,
+    )
+
+    return {sk[0]: sk[1] for sk in top_green_skills if sk[1]}
+
+
+def get_green_skill_matches(
+    extracted_skill_list: List[str],
+    similarities: np.array,
+    green_skills_taxonomy: pd.DataFrame(),
```
Comment (india-kerle) [inline dap_prinz_green_jobs/pipeline/green_measures/skills/green_skill_classifier.py]:
```suggestion
    green_skills_taxonomy: pd.DataFrame,
```

File: dap_prinz_green_jobs/pipeline/green_measures/skills/map_skills_utils.py [hunk 16]
```diff
@@ -0,0 +1,356 @@
+"""
+Utils to match skill embeddings to all of the ESCO taxonomy. This is a temporary measure since at the moment there
+is no functionality in ojd_daps_skills to take in already embedded skills and map them.
+
+As such, very little in this script is actually new code. It is mostly all copied from ojd_daps_skills, specifically,
+- ojd_daps_skills.pipeline.skill_ner_mapping.skill_ner_mapper.py
+- ojd_daps_skills.config.extract_skills_esco.yaml
+
+"""
```
Comment (india-kerle) [inline dap_prinz_green_jobs/pipeline/green_measures/skills/map_skills_utils.py:9]: I didn't review this btw!

File: dap_prinz_green_jobs/tests/test_skill_measures.py [hunk 17]
```diff
@@ -0,0 +1,86 @@
+import pytest
```
Comment (india-kerle) [inline dap_prinz_green_jobs/tests/test_skill_measures.py:1]: all tests passed with `9 passed, 45 warnings` 🤔 mostly `DeprecationWarning` warnings across numpy and scipy.

File: dap_prinz_green_jobs/utils/bert_vectorizer.py [hunk 18]
```diff
@@ -1,7 +1,39 @@
...
+
+import time
 import logging
+from tqdm import tqdm
+
+
+def get_embeddings(
+    sent_list: list, chunk_size: int = 1000, id_list: list = None
+) -> dict:
+    """
```
Comment (india-kerle) [inline dap_prinz_green_jobs/utils/bert_vectorizer.py:15]: this was a smart add

File: dap_prinz_green_jobs/utils/bert_vectorizer.py [hunk 19]
```diff
@@ -1,7 +1,39 @@
 from sentence_transformers import SentenceTransformer
-import time
 from ojd_daps_skills import logger
+import numpy as np
+
+from dap_prinz_green_jobs.utils.processing import list_chunks
```
Comment (india-kerle) [inline dap_prinz_green_jobs/utils/bert_vectorizer.py:5]: i forgot about the processing utils library ! can also use toolz's `partition` method to split a list into equally-sized chunks.

**Review by india-kerle — APPROVED** [review]
Thanks for the PR @lizgzil ! I tried reviewing this PR earlier in the week but came up against not being able to download your green skills classifier. It turned out it was because of scikit learn version issues - once solved, everything ran perfectly. We may want to think about dealing with requirements as part of the clean up of the repo in the next sprints.

really clear docstrings and comments in .py files which made following along with your code v smooth!

File: dap_prinz_green_jobs/pipeline/green_measures/green_measures.py (hunk 2 above)
Comment (lizgzil) [reply dap_prinz_green_jobs/pipeline/green_measures/green_measures.py]: good shout

File: dap_prinz_green_jobs/pipeline/green_measures/skills/README.md (hunk 5 above)
Comment (lizgzil) [reply dap_prinz_green_jobs/pipeline/green_measures/skills/README.md:69]: hmm annoying! well glad it was sortable. Do you think we should remove "ojd-daps-skills" from the requirements, but add to the README that if it's needed you should do a `pip install ojd-daps-skills`?

Comment (lizgzil) [reply dap_prinz_green_jobs/pipeline/green_measures/skills/README.md:69]: actually I will change the requirements to what Solomon suggested, i.e. 
```
numpy==1.24.4
pandas==1.3.5
scikit-learn==1.3.1
ojd-daps-skills @ git+https://github.com/nestauk/ojd_daps_skills.git@201_update_dependencies
```

File: dap_prinz_green_jobs/pipeline/green_measures/skills/green_skill_classifier.py (hunk 8 above)
Comment (lizgzil) [reply dap_prinz_green_jobs/pipeline/green_measures/skills/green_skill_classifier.py]: oo thanks!

File: dap_prinz_green_jobs/pipeline/green_measures/skills/green_skill_classifier.py (hunk 11 above)
Comment (lizgzil) [reply dap_prinz_green_jobs/pipeline/green_measures/skills/green_skill_classifier.py]: thanks! I'm so bad at these

File: dap_prinz_green_jobs/pipeline/green_measures/skills/green_skill_classifier.py (hunk 14 above)
Comment (lizgzil) [reply dap_prinz_green_jobs/pipeline/green_measures/skills/green_skill_classifier.py]: good idea

File: dap_prinz_green_jobs/tests/test_skill_measures.py (hunk 17 above)
Comment (lizgzil) [reply dap_prinz_green_jobs/tests/test_skill_measures.py:1]: I'm on 15 warnings. It's not clear to me what's causing them apart from package versions?

**Comment by lizgzil** [comment]
thanks for the review @india-kerle ! I've addressed your comments and made the fixes @sqr00t suggested in Slack to the requirements.txt.

**Comment by sqr00t** [comment]
~~heya, just one smaller change needed: removing awscli from requirements_dev.txt~~ Just pushed directly since it's a small change, easy to re-add if needed.

**Pull request merged**